File: 0274-h-index/0274-h-index.py
# Since this is not sorted, we could use bucket sort - single pass
class Solution:
    def hIndex(self, citations: List[int]) -> int:
        n = len(citations)
        buckets = [0] * (n + 1)
        count = 0
        
        for i in range(n):
            if citations[i] >= n:
                buckets[n] += 1
            else:
                buckets[citations[i]] += 1
        for i in range(n, -1, -1):
            count += buckets[i]
            if count >= i:
                return i
        return 0
        
        

File: 0274-h-index/0274-h-index.java
class Solution {
    public int hIndex(int[] citations) {
        int n = citations.length;
        int[] bucket = new int[n+1];

        for(int i =0;i <n; i++){
            if (citations[i] > n) {
                bucket[n] += 1;
            }
            else{
                bucket[citations[i]] += 1;
            }
            
        }
        int count = 0;
        for(int i=n; i >=0; i--){
            count += bucket[i];
            if (count >= i){
                return i;
            }
        }
    return 0;
    }
}